Added 1-Click Deploy tutorial (#160)
DefangLabs/defang-docs: docs/tutorials/adding-custom-one-click-deploy.md

---

> title: Adding a Custom 1-Click Deploy
> sidebar_position: 800
> description: Add a custom 1-Click Deploy button to deploy your own app. 
> ---
> 
> # Adding a Custom 1-Click Deploy

Reviewer: This title feels a little indirect. What do you think about something like this:

```suggestion
title: Using 1-Click Deploy With Your Project
sidebar_position: 800
description: Add a custom 1-Click Deploy button to deploy your own app. 
---

# Using 1-Click Deploy With Your Project
```

Author: I tried to make the title as different from the "Using 1-Click Deploy" tutorial as possible, so that it's clear for new users which one to look for first. Maybe "Adding Custom 1-Click Deploy to an App" will be better?

Reviewer: How about "Adding Custom 1-Click Deploy to _your_ App"?

Author: What about "Adding Custom 1-Click Deploy to Existing Apps"

Author: Or even just "Adding Custom 1-Click Deploy" - I like it, but would it be too vague?

This tutorial will show you how to add a custom Defang 1-Click Deploy button to deploy your own app to Defang Playground. 

:::info
This page focuses on adding 1-Click Deploy to any app that is not from [our list of samples](https://defang.io/#samples). We have a separate tutorial on [how to use 1-Click Deploy](/docs/tutorials/using-one-click-deploy) for Defang's pre-made samples, which is meant for new users. 
:::

## Step 1 - Have an App Ready

> The button we will be creating in this tutorial will allow any user to deploy your app to their Defang account. If your app is ready to be deployed with Defang (i.e. has a valid `compose.yaml` file), you can skip this step. 

Reviewer: ```suggestion
The button we will be creating in this tutorial will allow you to deploy your app to their Defang account. If your app is ready to be deployed with Defang (i.e. has a valid `compose.yaml` file), you can skip this step. 
```

### Option 2 - Manually Set Up a Template Repo

1. Create a new repository on GitHub. 
2. Go to the repository's "Settings", seen near the right on the top bar. 
3. Select the "Template repository" box to turn your repo into a template repo, as seen below:

![select-template-repo-box](/img/custom-one-click-tutorial/select-template-repo-box.png)

4. Go back to this repository's home page and follow the instructions in Option 1, skipping the first step.

## Step 4 - Get the Encoded URL

Now that you've created a new repository from a template repo, you will need the encoded version of the URL. This is necessary because it is used as a redirect for the 1-Click Deploy link. 

1. Copy the URL of your new repository's homepage (not the template repository's homepage). It should look something like the following:
```
https://github.com/<your-github-username>/<your-project-here>
```
2. You can go online and paste the URL from the step above into a URL encoder tool of your choice. You should end up with an encoded URL, similar to the one shown below:

```
https%3A%2F%2Fgithub.com%2F<your-github-username>%2F<your-project-here>
```

## Step 5 - Create the 1-Click Deploy Link

> Finally, you will need to format the 1-Click Deploy Link as `https://portal.defang.dev/redirect?url=` + your encoded URL. This ensures that the user can get [logged in](/docs/concepts/authentication/) to Defang before they get redirected to clone your app for deployment. 

Reviewer: We should be careful to write either directly to the user or indirectly to them, but not both. In this document it's mostly written to "you" the reader, so let's keep it consistent.

```suggestion
Finally, you will need to format the 1-Click Deploy Link as `https://portal.defang.dev/redirect?url=` + your encoded URL. This ensures that you can get [logged in](/docs/concepts/authentication/) to Defang before you get redirected to clone your app for deployment.
```

Reviewer: Ah, nevermind. Once I actually started to read this article and understand the headspace of a reader, I realize that this construction makes sense. Please disregard my previous suggestion. Sorry!

---
title: Adding a Custom 1-Click Deploy
sidebar_position: 800
description: Add a custom 1-Click Deploy button to deploy your own app. 
---

# Adding a Custom 1-Click Deploy

This tutorial will show you how to add a custom Defang 1-Click Deploy button to deploy your own app to Defang Playground. 

:::info
This page focuses on adding 1-Click Deploy to any app that is not from [our list of samples](https://defang.io/#samples). We have a separate tutorial on [how to use 1-Click Deploy](/docs/tutorials/using-one-click-deploy) for Defang's pre-made samples, which is meant for new users. 
:::

> <!-- 

Author: Oh yes ,that was a comment mainly for me to use while making the doc. I'll remove it.

```
defang generate
```

:::info
Make sure your `compose.yaml` file is in the root folder of your project.
:::

It may be helpful to have a `README.md` file, or a location to paste your 1-Click Deploy link. 

## Step 2 - Copy the GitHub Workflow
 
In your project, add a new folder named `.github`, then add a folder `workflows` inside it, and copy the entire GitHub Workflow `deploy.yaml` file from [here](https://github.com/DefangLabs/samples/blob/main/starter-sample/.github/workflows/deploy.yaml) into the `workflows` folder. 
 
If your app does not have any [configuration](/docs/concepts/configuration) needed (e.g. API keys, passwords, etc.), you can remove the corresponding lines as indicated in the file. 

You should have a `.yaml` file similar to the one below:
```yaml
name: Deploy

on:
  push:
    branches:
      - main

jobs:
  deploy:
    environment: playground
    runs-on: ubuntu-latest
    permissions:
      contents: read
      id-token: write

    steps:
    - name: Checkout Repo
      uses: actions/checkout@v4

    - name: Deploy
      uses: DefangLabs/defang-github-action@v1.1.3
```

## Step 3 - Use a Template GitHub Repository

> You have the option of using Defang's template repo or your own template GitHub Repo for this step. There is no difference in end result, but it may be easier to use Defang's template repo so that you do not have to create your own.

Reviewer 2: inconsistent capitalization of "template repo" versus GitHub Repo versus Template Repo below

### Option 1 - Use an Existing Template Repo

1. Click the link to [Defang's 1-Click Template Repo](https://github.com/DefangLabs/1-click-template), or open an existing template repo if you have one.  
2. Click the dropdown menu next to the green "Use this template" button near the top right corner, then click "Create a new repository":

> ![use-this-template](/img/custom-one-click-tutorial/use-this-template.png)

Reviewer 2: Step 3 missing? Also since the outer list is also numbered 1, 2, 3, 4, ... perhaps we should use a) b) c) for the sub-numbering here.

Author: good catch - I'll update it. the markdown format doesn't like open parentheses like ")" so I'll stick to the numbering system for now as it has auto-indent features.

4. You will be taken to a page similar to the one shown below. Then, enter a name for your repository and click the green "Create repository" button at the bottom.

![create-a-new-repo](/img/custom-one-click-tutorial/create-a-new-repo.png)

5. Now that you've created a new repository from a template repo, you can move onto the next step, which is getting the encoded URL.

### Option 2 - Manually Set Up a Template Repo

> 1. Create a new repository on GitHub. 

Reviewer 2: Similar comment - use a) b) c) sub-numbering

2. Go to the repository's "Settings", seen near the right on the top bar. 
3. Select the "Template Repository" box to turn your repo into a template repo, as seen below:

![select-template-repo-box](/img/custom-one-click-tutorial/select-template-repo-box.png)

4. Go back to this repository's home page and follow the instructions in Option 1, skipping the first step.

## Step 4 - Get the Encoded URL

Now that you've created a new repository from a template repo, you will need the encoded version of the URL. This is necessary because it is used as a redirect for the 1-Click Deploy link. 

> 1. Copy the URL of your new repository's homepage (not the template repository's homepage). It should look something like the following:

Reviewer 2: Similar comment on sub-numbering
